### ISSUE TITLE
Define a bi-directional data channel between opening and presenting browsing contexts

### ISSUE THREAD

**Issue by drott** [w3c/presentation-api#46]
http://w3c.github.io/presentation-api/#algorithm-post-message

Consider additional binary datatypes? https://bugzilla.mozilla.org/show_bug.cgi?id=1080474#c22

**Comment by mfoltzgoogle** [comment]
Candidates, similar interfaces, are found in the [WHATWG HTML spec's `MessageEvent` section] (https://html.spec.whatwg.org/multipage/comms.html#the-messageevent-interfaces) or [WebRTC spec for `RPCDataChannel`](http://w3c.github.io/webrtc-pc/#rtcdatachannel)

**Comment by mfoltzgoogle** [comment]
Mozilla currently supports [this interface for messaging](https://bugzilla.mozilla.org/attachment.cgi?id=8527264&action=diff#a/dom/webidl/PresentationSocketChannel.webidl_sec1):

```
interface PresentationSession : EventTarget {
  <snip>
  readonly attribute PresentationSocketChannel channel;
};
```

```
interface PresentationSocketChannel : EventTarget {
17	  [Throws]
18	  void send (DOMString data);
19	  [Throws]
20	  void send (Blob data);
21	  [Throws]
22	  void send (ArrayBuffer data);
23	  [Throws]
24	  void send (ArrayBufferView data);
25	
26	  void close ();
27	
28	  attribute PresentationBinaryType binaryType;
29	  readonly attribute unsigned long bufferedAmount;
30	
31	  const unsigned short CONNECTING = 0;
32	  const unsigned short OPEN = 1;
33	  const unsigned short CLOSING = 2;
34	  const unsigned short CLOSED = 3;
35	  readonly attribute unsigned short readyState;
36	
37	  attribute EventHandler onmessage;
38	  attribute EventHandler onopen;
39	  attribute EventHandler onerror;
40	  attribute EventHandler onclose;
41	};
```

I like that the API supports multiple data types, but it seems to assume the underlying connection technology has the same semantics as [WebSockets](http://www.w3.org/TR/websockets/).

My questions for the Mozilla folks:

(1) What does it mean for the PresentationSession to be connected but the associated channel to be CLOSED?  Similarly what does it mean for the session to be disconnected but the channel != CLOSED?

(2) What if the connection technology between the user agents is connection-less (i.e. based on UDP or a push messaging service)?  What would the readyState or state transitions mean?

I would instead propose a simpler API that allows flexible data passing without making assumptions about the connection technology:

```
interface PresentationSession : EventTarget {
	  [Throws]
	  void send (DOMString data);

	  [Throws]
	  void send (Blob data);

	  [Throws]
	  void send (ArrayBuffer data);

	  [Throws]
	  void send (ArrayBufferView data);

	  attribute EventHandler onmessage;
};
```

**Comment by schien** [comment]
The latest WIP we have in hand is pretty much like your proposal. We are not going to provide an extra "Socket" interface because of the confusing during state transition. In addition, we are considering about always using a connection technology which is reliable and ordered, therefore, the behavior of message passing in 1-UA and 2-UA mode will be the same.

**Comment by anssiko** [comment]
I'm hearing the proposals are starting to converge, which is encouraging.

@drott Please indicate whether you'd prefer to take the first stab at landing this, or whether you'd like other people in this issue to submit a PR.

Thanks @mfoltzgoogle and @schien for making progress on this issue!

**Comment by mfoltzgoogle** [comment]
Great! Let me know if you would like help with a PR.

**Comment by drott** [comment]
I'll write a first version - would be great if @mfoltzgoogle  and @schien could review it. Thanks.

**Comment by drott** [comment]
I do not intend to merge this yet - it needs a bit more work, but I would appreciate an initial review to see whether we're on the same page with this.

**Comment by avayvod** [comment]
I think we should have left the name send() since it's aligned with RTCDataChannel send(), not with various existing postMessage() APIs - using postMessage in this sense is confusing.

**Comment by mfoltzgoogle** [comment]
I went ahead and merged and pushed Anton's pull request to update the specification to use `send()` in place of `postMessage()`.

**Comment by anssiko** [comment]
All - I suggest we review this issue at F2F since communication between contexts is a core part of the API. We also seem to have multiple work-in-progress implementations, so my expectation is implementers (ping @mfoltzgoogle @schien) would be able to share some concrete implementation feedback at F2F regarding this feature.

* API http://w3c.github.io/presentation-api/#presentationsession:
```
// Communication
  attribute BinaryType binaryType;
  EventHandler onmessage;
  void send (DOMString message);
  void send (Blob data);
  void send (ArrayBuffer data);
  void send (ArrayBufferView data);
```

* Algorithms and prose:
 * http://w3c.github.io/presentation-api/#sending-a-message-through-presentationsession
 * http://w3c.github.io/presentation-api/#receiving-a-message-through-presentationsession

**Comment by anssiko** [comment]
Related https://github.com/w3c/presentation-api/issues/63

**Comment by matt-hammond-bbc** [comment]
If the API is designed to allow connectionless implementations, then could there be a UDP based implementation that does not guarantee message deliver? This would have a big impact on how the messaging channel must be used.

Currently the text only states that the sequence should be preserved. My proposal would be to also explicitly state whether message delivery can be assumed to be reliable/unreliable.

**Comment by tidoust** [comment]
PROPOSED RESOLUTION: adopt the "send"-like interface and clarify in the spec that the data channel is an in-order reliable message-based communication channel

**Comment by tidoust** [comment]
Relevant discussion during first day of the F2F in Berlin:
http://www.w3.org/2015/05/19-webscreens-minutes.html#item07

**Comment by anssiko** [comment]
ACTION: @tidoust to look at how WebSockets and how WebRTC data channels deal with empty messages [recorded in http://www.w3.org/2015/05/19-webscreens-minutes.html#action10]

**Comment by tidoust** [comment]
During the F2F, we wondered about how WebSockets and WebRTC handled messages consisting of and empty string. I investigated and totally agree with @mwatson2 that the algorithms described in both specs do not treat empty strings in any special way, and thus empty strings should be treated as any other string message, following the exact same steps.

In other words, calling `send('')` should send an empty string over the communication channel and trigger a `message` event on the receiving side. This is exactly what happens in all WebSockets implementations that I tested.

However, as reported by @avayvod during the F2F, Chrome and Firefox silently drop empty string messages sent over a WebRTC data channel for the time being. I believe that is a bug in the way they interpret the specification and I submitted a test case for the WebRTC test suite that specifically checks that point, see:

  https://github.com/w3c/web-platform-tests/pull/1897

I do not think that we need to change anything in the Presentation API for that.
